File: 2022/day-05/day-05.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" --- Advent of Code 2022 - Day 5: Supply Stacks --- """

import re
import sys


# initialize the program and return stacks as array and procedures as lines
def init() -> tuple:
    with open("input.txt", "r") as f:
        crates, procedures = f.read().split("\n\n")
        stacks = create_stacks(crates)
        return stacks, procedures.split("\n")


# transform the given string stacks into 2d array stacks
def create_stacks(crates: str) -> list:
    *crates, num_stacks = crates.split("\n")
    num_stacks = len(num_stacks.split())

    # creates an empty 2d array of length 9
    stacks = [[] for _ in range(num_stacks)]

    # put crates into 2d array at stackIdx from bottom to top
    crates.reverse()
    for line in crates:
        for stackIdx, i in enumerate(range(1, len(crates[0]), 4)):
            if line[i].strip():
                stacks[stackIdx].append(line[i])

    return stacks


# for every instruction find the 3 important numbers: #, from, to
def parse_instructions(instruction: str) -> tuple:
    crates_to_move, to_stack, from_stack = map(int, re.findall(r"[0-9]*[0-9]+", instruction))
    return crates_to_move, to_stack-1, from_stack-1


# solution part 1
def part_1():
    stacks, procedures = init()
    for procedure in procedures:
        if procedure.strip():
            crates_to_move, from_stack, to_stack = parse_instructions(procedure)

            # move one crate at a time
            for _ in range(crates_to_move):
                stacks[to_stack].append(stacks[from_stack].pop())

    # print the last element of each list in a list of lists
    [print(stack[-1], end="") for stack in stacks]
    print()


# solution part 2
def part_2():
    stacks, procedures = init()
    for procedure in procedures:
        if procedure.strip():
            crates_to_move, from_stack, to_stack = parse_instructions(procedure)

            # move crates at once
            stacks[to_stack].extend(stacks[from_stack][-crates_to_move:])
            stacks[from_stack] = stacks[from_stack][:-crates_to_move]

    # print the last element of each list in a list of lists
    [print(stack[-1], end="") for stack in stacks]
    print()


# main entry point
if len(sys.argv) > 1 and sys.argv[1] == "-p1":
    part_1()
elif len(sys.argv) > 1 and sys.argv[1] == "-p2":
    part_2()
else:
    print("Usage: python", sys.argv[0], "[-p1|-p2]")
    sys.exit(1)

File: 2022/day-05/day-05.c
/* --- Advent of Code 2023 - Day 5: Supply Stack --- */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LENGTH 256
#define MAX_STACKS 32
#define MAX_HEIGHT 32

/* solution part-1 */
int part_1(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];

    /* store stacks in 2d array and keep track of their height */
    int no_stacks = 0;
    int stacks[MAX_STACKS][MAX_HEIGHT];
    int stack_height[MAX_STACKS];
    for (int i = 0; i <= MAX_STACKS; i++) stack_height[i] = 0;

    /* file opened successfully */
    if (fp != NULL) {

        /* create stacks */
        while (fgets(buffer, MAX_LENGTH, fp) && buffer [1] != '1') {
            no_stacks = strlen(buffer) / 4;

            /* fill stacks with crates */
            for (int i = 0; i < no_stacks; i++) {
                if (buffer[i * 4] == '[') {
                    stacks[i + 1][stack_height[i + 1]++] = buffer[i * 4 + 1];
                }
            }
        }

        /* reverse stacks */
        for (int i = 1; i <= no_stacks; i++) {
            for (int j = 0; j < stack_height[i] / 2; j++) {
                char tmp = stacks[i][j];
                stacks[i][j] = stacks[i][stack_height[i] - j - 1];
                stacks[i][stack_height[i] - j - 1] = tmp;
            }
        }

        /* skip empty line */
        fgets(buffer, MAX_LENGTH, fp);

        /* move one crate at a time */
        while (fgets(buffer, MAX_LENGTH, fp)) {
            int crates_to_move, dest, src;
            if (sscanf(buffer, "move %d from %d to %d", &crates_to_move, &src, &dest) != 3) {
                fprintf(stderr, "Error parsing line '%.*s'\n", (int) strlen(buffer), buffer);
                exit(1);
            }

            for (int i = 0; i < crates_to_move; i++) {
                stacks[dest][stack_height[dest]++] = stacks[src][--stack_height[src]];
            }
        }

        /* print crates at the top of each stack */
        for (int i = 1; i <= no_stacks; i++) {
            printf("%c", stacks[i][stack_height[i] - 1]);
        }
        printf("\n");

    /* file not found */
    } else {
        fprintf(stderr, "Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* solution part-2 */
int part_2(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];

    /* store stacks in 2d array and keep track of their height */
    int no_stacks = 0;
    int stacks[MAX_STACKS][MAX_HEIGHT];
    int stack_height[MAX_STACKS];
    for (int i = 0; i <= MAX_STACKS; i++) stack_height[i] = 0;

    /* file opened successfully */
    if (fp != NULL) {
        /* create stacks */
        while (fgets(buffer, MAX_LENGTH, fp) && buffer[1] != '1') {
            no_stacks = strlen(buffer) / 4;

            /* fill stacks with crates */
            for (int i = 0; i < no_stacks; i++) {
                if (buffer[i * 4] == '[') {
                    stacks[i + 1][stack_height[i + 1]++] = buffer[i * 4 + 1];
                }
            }
        }

        /* reverse stacks */
        for (int i = 1; i <= no_stacks; i++) {
            for (int j = 0; j < stack_height[i] / 2; j++) {
                char tmp = stacks[i][j];
                stacks[i][j] = stacks[i][stack_height[i] - j - 1];
                stacks[i][stack_height[i] - j - 1] = tmp;
            }
        }

        /* skip empty line */
        fgets(buffer, MAX_LENGTH, fp);

        /* move crates at once */
        while (fgets(buffer, MAX_LENGTH, fp)) {
            int crates_to_move, dest, src;
            sscanf(buffer, "move %d from %d to %d", &crates_to_move, &src, &dest);

            for (int i = 0; i < crates_to_move; i++) {
                stacks[dest][stack_height[dest]++] = stacks[src][stack_height[src] - crates_to_move + i];
            }
            stack_height[src] -= crates_to_move;
        }

        /* print crates at the top of each stack */
        for (int i = 1; i <= no_stacks; i++) {
            printf("%c", stacks[i][stack_height[i] - 1]);
        }
        printf("\n");

    /* file not found */
    } else {
        fprintf(stderr, "Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* main function */
int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s -p[1|2] <input-file>\n", argv[0]);
        return 1;
    }

    if (!strcmp(argv[1], "-p1")) {
        part_1(argv[2]);
    } else if (!strcmp(argv[1], "-p2")) {
        part_2(argv[2]);
    } else {
        fprintf(stderr, "Invalid argument: '%s'\n", argv[1]);
        return 1;
    }

    return 0;
}
